nsis/src/nsis/NsisLoader.java

```diff
@@ -90,46 +95,68 @@ protected void load(ByteProvider provider, LoadSpec loadSpec,
-				return;
-			}
-
-			// TODO we probably don't need both types of readers (input
-			// stream and binary reader)
-			InputStream inputStream;
-			inputStream = provider.getInputStream(nsis_header_offset);
-			Memory mem = program.getMemory();
+			long scriptHeaderOffset = ne.getHeaderOffset();
 
 			BinaryReader binary_reader = new BinaryReader(provider,
 					/* isLittleEndian= */ true);
-			binary_reader.setPointerIndex(nsis_header_offset);
-
-			ghidra.program.model.address.Address script_header_start = program
-					.getAddressFactory().getDefaultAddressSpace()
-					.getAddress(0x0);
+			binary_reader.setPointerIndex(scriptHeaderOffset);
 
-			MemoryBlock new_block = mem.createInitializedBlock(".script_header",
-					script_header_start, inputStream,
-					ne.getInflatedHeaderSize(), monitor, false);
-			new_block.setRead(true);
-			new_block.setWrite(true);
-			new_block.setExecute(true);
+			Address scriptHeaderAddress = program.getAddressFactory()
+					.getDefaultAddressSpace().getAddress(scriptHeaderOffset);
+			FileBytes fileBytes = MemoryBlockUtils.createFileBytes(program,
+					provider, scriptHeaderOffset, ne.getArchiveSize(), monitor);
 
-			createData(program, program.getListing(), script_header_start,
-					ne.getHeaderDataType());
-			processBlockHeaders(program, monitor, binary_reader,
-					nsis_header_offset);
+			initScriptHeader(fileBytes, scriptHeaderAddress,
+					fileBytes.getSize(), program, ne.getHeaderDataType());
+			initBlockHeaders(program, binary_reader, scriptHeaderAddress);
 
-			System.out.printf("Done initializing block headers\n");
 		} catch (Exception e) {
-			System.out.print(e.getMessage());
-			log.appendException(e);
+			throw new IOException(e); // Ghidra handles the thrown exception
 		}
 	}
 
-	public Data createData(Program program, Listing listing, Address address,
+	/**
+	 * Initializes the script header and adds it to the "Program Trees" view in
+	 * Ghidra.
+	 * 
+	 * @param fileBytes            object that starts at the NSIS magic bytes
+	 * @param scriptHeaderAddress, the address at which the nsis script header
+	 *                             starts
+	 * @param size                 of the header
+	 * @param program              object
+	 * @param dataType             of the script header
+	 * @throws MemoryConflictException
+	 * @throws AddressOverflowException
+	 * @throws CancelledException
+	 * @throws DuplicateNameException
+	 * @throws LockException
+	 */
+	private void initScriptHeader(FileBytes fileBytes,
+			Address scriptHeaderAddress, long size, Program program,
+			DataType dataType)
+			throws MemoryConflictException, AddressOverflowException,
+			CancelledException, DuplicateNameException, LockException {
+		Memory memory = program.getMemory();
+		MemoryBlock new_block = memory.createInitializedBlock(".script_header",
+				scriptHeaderAddress, fileBytes, 0, size, false);
+		new_block.setRead(true);
+		new_block.setWrite(true);
+		new_block.setExecute(true);
```

Comment:
For the header block I would be tempted to set it to, normally we wouldn't write or execute anything in the header.

```java
new_block.setRead(true);
new_block.setWrite(false);
new_block.setExecute(false);
```